eth/downloader/statesync.go

```diff
@@ -0,0 +1,245 @@
+			s.requestDone(req)
+		}
+	}()
+
+	newPeer := make(chan *peer, 200)
+	peerSub := s.d.peers.SubscribeNewPeers(newPeer)
+	defer peerSub.Unsubscribe()
+
+	for {
+		if s.sched.Pending() == 0 {
+			return nil
+		}
+		s.assignTasks()
+
+		select {
+		case packI := <-deliver:
+			pack := packI.(*statePack)
+			req := s.fetching[pack.peerId]
+			if req == nil {
+				log.Debug("Unrequested node data", "peer", pack.peerId, "len", len(pack.states))
+				continue
+			}
+			s.requestDone(req)
+			req.peer.SetNodeDataIdle(pack.Items())
+			if pack.Items() == 0 {
+				log.Trace("Empty node data pack", "peer", pack.peerId)
+				for _, item := range req.items {
+					req.peer.MarkLacking(item)
+				}
+				continue
+			}
+			procStart := time.Now()
+			n, err := s.process(req, pack)
+			if err != nil {
+				log.Warn("Node data write error", "err", err)
+			}
+			s.updateStats(n, time.Since(procStart))
+
+		case req := <-s.timeout:
+			if s.fetching[req.peer.id] != req {
+				continue // ignore old timeout
+			}
+			log.Debug("Node data timeout", "peer", req.peer.id, "items", len(req.items))
+			s.requestDone(req)
+			s.retry = append(s.retry, req.items...)
+			for _, hash := range req.items {
+				req.peer.MarkLacking(hash)
+			}
+			req.peer.SetNodeDataIdle(0)
+
+		case <-newPeer:
+			// assign new tasks
+		case <-s.quit:
+			return errCancelStateFetch // aborted through s.cancel()
+		case <-s.d.cancelCh:
+			return errCancelStateFetch // downloader was canceled
+		}
+	}
+}
+
+func (s *stateSync) requestDone(req *stateReq) {
+	delete(s.fetching, req.peer.id)
+	close(req.cancelTimeout)
+}
+
+func (s *stateSync) assignTasks() {
+	peers, _ := s.d.peers.NodeDataIdlePeers()
+	for len(peers) > 0 && (s.sched.QueueLen() > 0 || len(s.retry) > 0) {
+		if peer := peers[0]; s.fetching[peer.id] == nil {
+			rtt := s.d.requestRTT()
+			items := s.popTasks(peer.NodeDataCapacity(rtt))
+			s.fetching[peer.id] = s.spawnFetch(peer, items, rtt)
+		}
+		peers = peers[1:]
+	}
+}
+
+func (s *stateSync) popTasks(n int) []common.Hash {
+	items := make([]common.Hash, 0, n)
+	// add from retry list
+	for ; n >= 0 && len(s.retry) > 0; n-- {
+		items = append(items, s.retry[len(s.retry)-1])
+		s.retry = s.retry[:len(s.retry)-1]
+	}
+	// add from the scheduler
+	return append(items, s.sched.Missing(n)...)
+}
+
+func (s *stateSync) spawnFetch(p *peer, nodes []common.Hash, timeout time.Duration) *stateReq {
+	log.Trace("Requesting node data", "peer", p.id, "items", len(nodes), "timeout", timeout)
+	req := &stateReq{p, nodes, make(chan struct{})}
+	time.AfterFunc(timeout, func() {
+		select {
+		case s.timeout <- req:
+		case <-req.cancelTimeout:
+		}
+	})
+	p.FetchNodeData(nodes)
+	return req
+}
+
+func (s *stateSync) process(req *stateReq, pack *statePack) (int, error) {
+	var (
+		procErr     error
+		nproc       int
+		anyProgress bool
+		batch       = s.db.NewBatch()
+	)
+	for _, blob := range pack.states {
+		hash := common.BytesToHash(crypto.Keccak256(blob))
```

Comment:
You probably want to reuse a hashes throughout the for loop. There are almost 400 states per request, so there's no reason to do so many hasher allocs and deallocs.